_posts/pl/2021-12-XX-transactions.md

```diff
@@ -0,0 +1,217 @@
+			return id;
+		});
+	}
+}
+```
+**Zalety**:
+- nie trzeba zaprzątać sobie głowy otwieraniem i zamykaniem połączenia z bazą danych poprzez stosowanie bloku _try-catch-finally_. Zamiast tego wykorzystywany jest mechanizm callback’ów (`Transaction Callbacks`).
+- nie trzeba przechwytywać wyjątków _SQLException_, ponieważ Spring konwertuje je na wyjątki _RuntimeException_.
+- lepsza integracja z ekosystemem Springa. _TransactionTemplate_ wykorzystuje _TransactionManager_ do konfiguracji połączenie z bazą danych. Ich implementacja wiążę się z koniecznością utworzenia odpowiednich beanów, jednak później nie musimy się już martwić o zarządzanie nimi.
+
+## Deklaratywne zarządzanie transakcją
+O ile powyższa metoda jest znacznym usprawnieniem i ułatwieniem w zarządzaniu transakcjami, to  Spring daje nam nieco więcej, a wszystko za sprawą adnotacji **`@Transactional`**. Dzięki niej, każda publiczna metoda, która zostanie oznaczona tą adnotacją, zostanie wykonana wewnątrz transakcji.
+Mając już wiedzę w jaki sposób wygląda zarządzanie transakcją w JDBC, zobaczmy jak łatwo przekształcić tak, aby wykorzystywał dobrodziejstwo frameworka Spring:
+
+Na początek zobaczmy jak wyglądałoby dodanie nowej książki do bazy przy użyciu JDBC:
+```java
+public class BookService {
+
+    public Long addBook(Book book) {
+        Connection connection = dataSource.getConnection();
+        try (connection) {
+            connection.setAutoCommit(false);
+
+	// dodaj książkę do bazy i zwróć wygenerowany identyfikator 
+            // bookDao.save(book);
+
+            connection.commit();
+        } catch (SQLException e) {
+            connection.rollback();
+        }
+    }
+}
+```
+A teraz przyjrzyjmy się użyciu adnotacji _@Transactional_ w Springu:
+```java
+public class BookService {
+	@Transactional
+	public Long addBook(Book book) {
+		// dodaj książkę do bazy i zwróć wygenerowany identyfikator 
+		// bookDao.save(book);
+		return id;
+	}
+}
+```
+Wygląda znacznie prościej prawda? Aby skorzystać z adnotacji _@Transactional_, konieczne jest jedynie zdefiniowanie menadżera transakcji w konfiguracji Spring’a oraz dodanie adnotacji _@EnableTransactionManagement_ (w Spring Boot nie jest to konieczne).
+```java
+@Configuration
+@EnableTransactionManagement
+public class MyConfig {
+
+    @Bean
+    public PlatformTransactionManager transactionManager() {
+        return transactionManager;
+    }
+}
+```
+
+## Jak działa adnotacja @Transactional
+Spring pozwala na wykorzystanie usługi BookService w każdym innym beanie, który tego wymaga. Stosując adnotację _@Transactional_ i odwołując się do metody, która jest nią opatrzona, Spring nie odwołuje się bezpośrednio do tej metody, ale tworzy tzw. **proxy transakcyjne**. Dzieje się to przy pomocy biblioteki _Cglib_ i metody zwanej _proxy-through-sublcassing_. Zadaniem tego proxy jest zawsze:
+- otwieranie i zamykanie połączeń/transakcji z bazą danych
+- delegowanie zadania do rzeczywistej usługi, jak nasze _BookService_
```

Comment:
Pisałeś wcześniej o poziomach izolacji, więc warto tutaj to wykorzytać i opisać, jak ma się to do @Transactional. Myśle o DEFAULT i pułapce w jaką można tutaj wpaść

Reply:
napisałem o DEFAULT